ts/metrics/metric_cache_abstract.py

```diff
@@ -0,0 +1,462 @@
+
+
+class MetricCacheAbstract(metaclass=abc.ABCMeta):
+    def __init__(self, request_ids, model_name, file):
+        """
+        Constructor for MetricsCaching class
+
+            MetricsCaching class will hold Metric objects in a cache that can be edited and read from.
+            The class also has the ability to parse a yaml file into Metrics objects
+
+        Parameters
+        ----------
+        file: str
+            Name of file to be parsed
+
+        """
+        self.cache = {}
+        self.store = []
+        self.request_ids = request_ids
+        self.model_name = model_name
+        self.file = file
+
+    @staticmethod
+    def _check_matching_dims(key, dims) -> bool:
+        """
+        Check to see if key value string pair already in list of dim strings
+        """
+        is_present = False
+        for dim in dims:
+            key_dim = dim.split(":")[0]
+            if key == key_dim:
+                is_present = True
+        return is_present
+
+    def _add_or_update(
+        self,
+        name: str,
+        value: int or float,
+        req_id,
+        unit: str,
+        metric_type: MetricTypes = None,
+        dimensions: list = None,
+    ):
+        """
+        Add a metric key value pair
+
+        Parameters
+        ----------
+        name : str
+            metric name
+        value: int, float
+            value of metric
+        req_id: str
+            request id
+        unit: str
+            unit of metric
+        value: int, float, str
+            value of metric
+        metric_type: str, optional
+            indicates type of metric operation if it is defined, creates default metric type if not defined
+        dimensions: list, optional
+            list of Dimension objects
+        """
+        # IF req_id is none error Metric
+        if dimensions is None or not dimensions:
+            dimensions = []
+        if not isinstance(dimensions, list):
+            raise merrors.MetricsCacheValueError(
+                "Please provide a list of Dimension objects."
+            )
+
+        # using a copy to not alter the original dimensions list when Metrics are added
+        dimensions = dimensions.copy()
+        # don't add duplicate Dimension objects
+        pre_dims_str = [str(d) for d in dimensions]
+
+        if req_id is None:
+            if not self._check_matching_dims("Level", pre_dims_str):
+                dimensions.append(Dimension("Level", "Error"))
+        else:
+            if not self._check_matching_dims("ModelName", pre_dims_str):
+                dimensions.append(Dimension("ModelName", self.model_name))
+            if not self._check_matching_dims("Level", pre_dims_str):
+                dimensions.append(Dimension("Level", "Model"))
+
+        dims_str = ",".join([str(d) for d in dimensions])
+        # convert metric enum into string
+
+        # Cache the metric with a unique key
+        metric_key = f"[{metric_type.value}]-[{name}]-[{dims_str}]"
```

Comment:
I think we should have keys that are objects instead of strings. a python dataclass would work great for this